pct.sh

```diff
@@ -18,23 +18,99 @@ if [[ -n ${EXTRA_MAVEN_PROPERTIES-} ]]; then
+
+#
+# The overrideWar option is available in HPI Plugin 3.29 or later, but many plugins under test
+# still use an older plugin parent POM and therefore an older HPI plugin version. As a temporary
+# workaround, we override the HPI plugin version to the latest version.
+#
+# TODO When all plugins in the managed set are using a plugin parent POM with HPI Plugin 3.29 or
+# later (i.e., plugin parent POM 4.44 or later), this can be deleted.
+#
+MAVEN_PROPERTIES+=:hpi-plugin.version=3.30
+
+#
+# Define the excludes for upper bounds checking. We define these excludes in a separate file and
+# pass it in via -mavenPropertiesFile rather than using -mavenProperties because -mavenProperties
+# uses a colon as the separator and these values contain colons.
+#
+if [[ $LINE == '2.319.x' ]]; then
+	#
+	# com.sun.mail needs to be excluded because it is still provided by core on this line (using
+	# pre-Jakarta imports) and we do not want it to be upgraded to a version that uses Jakarta
+	# imports (which is not a realistic test scenario) just because the Jakarta Mail API plugin
+	# happens to be on the class path and triggers an upper bounds violation.
+	#
+	# javax.servlet:servlet-api comes from core at version 0, which is an intentional trick to
+	# prevent this library from being used, and we do not want it to be upgraded to a nonzero
+	# version (which is not a realistic test scenario) just because it happens to be on the
+	# class path of some plugin and triggers an upper bounds violation. JENKINS-68696 tracks the
+	# removal of this trick.
+	#
+	echo upperBoundsExcludes=com.sun.mail:jakarta.mail,javax.servlet:servlet-api >"$(pwd)/maven.properties"
+else
+	#
+	# javax.servlet:servlet-api comes from core at version 0, which is an intentional trick to
+	# prevent this library from being used, and we do not want it to be upgraded to a nonzero
+	# version (which is not a realistic test scenario) just because it happens to be on the
+	# class path of some plugin and triggers an upper bounds violation. JENKINS-68696 tracks the
+	# removal of this trick.
+	#
+	echo upperBoundsExcludes=javax.servlet:servlet-api >"$(pwd)/maven.properties"
+fi
+
+#
+# Apache HttpComponents Client 4.x API, Oracle Java SE Development Kit Installer, JSch, OkHttp, and
+# Plain Credentials use an older plugin parent POM and therefore an older test harness that predates
+# compatibility with the removal of JNR in recent cores in jenkinsci/jenkins-test-harness#350. As a
+# temporary workaround, we override the test harness to a recent version. Note that we cannot use a
+# test harness newer than 1812.v6d4e97d91fd8, because later releases of the test harness require
+# changes to the plugin parent POM for JUnit 5 support.
+#
+# TODO When these plugins are using a plugin parent POM with test harness 1657.vf8a824e79147 or
+# later (i.e., plugin parent POM 4.32 or later), this can be deleted.
```

Comment:
Taking on a small amount of technical debt here, which I felt was reasonable in the short term because these five plugins are currently unmaintained. I will try and nag the maintainers a little more to get these plugins updated before I give up and adopt them myself. Help would certainly be appreciated if anyone is interested. I feel that this technical debt could be reasonably dealt with in the next couple of months, which is why I am choosing to proceed with this PR regardless.

Reply:
I guess you mean that this could become a problem when plugins start to rely on JTH features  (rather more likely than `maven-hpi-plugins` which I mentioned elsewhere), so yeah we will need to deal with this in the short-to-medium term.

I did not see anything in https://github.com/jenkinsci/plain-credentials-plugin/pulls and it looks like it is not yet set up with Dependabot or even JEP-305, oops. Will make a note to refresh it soon. I probably have rights to some of the others too, do not recall.

Reply:
> I did not see anything in https://github.com/jenkinsci/plain-credentials-plugin/pulls

Did you expect to? If you did, should I have expected to see a PR from you in https://github.com/jenkinsci/git-server-plugin/pulls after detaching `instance-identity`? :grinning: Come on Jesse, at least I am being upfront about taking a shortcut here. :rofl: